cpp/src/fil/fil.cu

```diff
@@ -625,13 +604,109 @@ int max_depth(const tl::ModelImpl<T, L>& model)
+          ASSERT(max_matching_cat <= MAX_PRECISE_INT_FLOAT,
+                 "FIL cannot infer on "
+                 "more than %d matching categories",
+                 MAX_PRECISE_INT_FLOAT);
+        } else {
+          max_matching_cat = -1;
+        }
+        int* max_matching_res = &res[tree.SplitIndex(node_id)];
+        *max_matching_res     = std::max(*max_matching_res, max_matching_cat);
+      }
+      stack.push(tree.LeftChild(node_id));
+      node_id = tree.RightChild(node_id);
+    }
+  }
+  return res;
+}
+
+// constructs a vector of size max_fid (number of features, or columns) from a Treelite tree,
+// where each feature has a maximum matching category and number of categorical nodes
+template <typename T, typename L>
+inline std::size_t bit_pool_size(const tl::Tree<T, L>& tree, cat_sets_owner& cat_sets)
+{
+  std::size_t size = 0;
+  std::stack<int> stack;
+  stack.push(tree_root(tree));
+  while (!stack.empty()) {
+    int node_id = stack.top();
+    stack.pop();
+    while (!tree.IsLeaf(node_id)) {
+      if (tree.SplitType(node_id) == tl::SplitFeatureType::kCategorical &&
+          tree.HasMatchingCategories(node_id)) {
+        int fid = tree.SplitIndex(node_id);
+        size += cat_sets.accessor().sizeof_mask(fid);
+        ++cat_sets.n_nodes[fid];
+      }
+      stack.push(tree.LeftChild(node_id));
+      node_id = tree.RightChild(node_id);
+    }
+  }
+  return size;
+}
+
+void vec_max(std::vector<int>& dst, const std::vector<int>& extra)
+{
+  std::transform(dst.begin(), dst.end(), extra.begin(), dst.begin(), [](int a, int b) {
+    return std::max(a, b);
+  });
+}
+
+template <typename T, typename L>
+cat_sets_owner allocate_cat_sets_owner(const tl::ModelImpl<T, L>& model)
+{
+#pragma omp declare reduction(vec_max_red : std::vector<int> \
+      : vec_max(omp_out, omp_in))                              \
+    initializer(omp_priv = omp_orig)
+  const auto& trees = model.trees;
+  cat_sets_owner cat_sets(model.num_feature, trees.size());
+  std::vector<int>& max_matching = cat_sets.max_matching;
+#pragma omp parallel for reduction(vec_max_red : max_matching)
+  for (size_t i = 0; i < trees.size(); ++i) {
+    vec_max(max_matching, max_matching_cat(trees[i], model.num_feature));
+  }
+#pragma omp parallel for
+  for (size_t i = 0; i < trees.size(); ++i) {
+    cat_sets.bit_pool_sizes[i] = bit_pool_size(trees[i], cat_sets);
+  }
+  cat_sets.initialize_from_bit_pool_sizes();
+  return cat_sets;
+}
+
+void adjust_threshold(
   float* pthreshold, int* tl_left, int* tl_right, bool* default_left, tl::Operator comparison_op)
 {
   // in treelite (take left node if val [op] threshold),
   // the meaning of the condition is reversed compared to FIL;
   // thus, "<" in treelite corresonds to comparison ">=" used by FIL
   // https://github.com/dmlc/treelite/blob/master/include/treelite/tree.h#L243
+  if (isnan(*pthreshold)) {
+    std::swap(*tl_left, *tl_right);
+    *default_left = !*default_left;
+    return;
+  }
```

Comment:
RAPIDS haven't updated to the version of treelite that supports empty categorical nodes, but this workaround works just fine.

Reply:
#4220 actually upgrades Treelite to 2.1.0 which supports empty categorical nodes.

Reply:
as soon as it's in, removing the export WAR be a quick change to test